### PR TITLE
Account list

### PR DESCRIPTION
**Pull request by Romulus10**
Adds account creation, balance/account querying functions.

### CODE REVIEW

File: Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt [hunk 1]
```diff
@@ -14,13 +17,14 @@ class BlockChain(var blockList: ArrayList<Block> = arrayListOf()) {
...
     }
 
-    fun addTransactionToQueue(transaction: Transaction){
+    fun addTransactionToQueue(transaction: Transaction) {
+        interpreter.runContract(transaction.data)
         transactionQueue.addTransaction(transaction)
     }
 
 
     //TODO replace empty validator/signature values
```
Comment (Romulus10) [inline Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt]: If I'm thinking properly, the validator should take the **public** key of the signer, and the signature is just the signature they applied. We have everything here for this now.

File: Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt [hunk 2]
```diff
@@ -30,32 +34,49 @@ class BlockChain(var blockList: ArrayList<Block> = arrayListOf()) {
...
     }
 
+    fun getBalanceFromAddress(address: String): Int {
+        return interpreter.accountList.filter {
+            it.address == address
+        }[0].weight
+    }
+
+    fun listKnownAddresses(): Array<Account> {
+        return interpreter.accountList.toArray() as Array<Account>
```
Comment (Romulus10) [inline Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt]: This has poor type safety. I couldn't get it to agree with me, I'll hack at it again, but if you can come up with anything to get rid of the warning feel free to propose a change.

File: Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt [hunk 3]
```diff
@@ -27,7 +32,7 @@ class Interpreter {
                     memory[instructions[index]?.get(1)!!.toInt()] = memory[instructions[index]?.get(2)!!.toInt()]
                 }
                 "2" -> { //trn
-                    // TODO Write an account value transfer function.
+                    transferAccountValue(accountList.toArray() as Array<Account?>, instructions[index]?.get(2), instructions[index]?.get(3), instructions[index]?.get(1)!!.toInt())
```
Comment (Romulus10) [inline Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: This has type safety issues. Refer to the other times this "as Array<*>" syntax is used.

File: Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt (hunk 1 above)
Comment (cvennel) [reply Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt]: @Romulus10 I wasn't sure if validator would be the public key or the human readable account name so we can get the associated public key. Passing public key was what I was leaning towards.

File: Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt (hunk 2 above)
Comment (cvennel) [reply Simulator/src/main/kotlin/edu/drexel/se575/BlockChain.kt]: Pushed an edit for this.

File: Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt (hunk 3 above)
Comment (cvennel) [reply Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: Could replace with 
`accountList.toArray().filterIsInstance<Account?>().toTypedArray()`

Comment (cvennel) [reply Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: Or if you want to ensure the cast went correctly:
`accountList.toArray().filterIsInstance<Account?>().toTypedArray().apply { if (size != accountList.size) throw Exception() }`

Comment (Romulus10) [reply Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: Well the worst they would be is null, their type is always defined explicitly. We solve that by using the ? nullable, so if I'm thinking correctly that should always be type-safe.

Comment (Romulus10) [reply Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: > Or if you want to ensure the cast went correctly:
> `accountList.toArray().filterIsInstance<Account?>().toTypedArray().apply { if (size != accountList.size) throw Exception() }`

That's even better. We'll just want to keep track of all the unhandled exceptions, that could get ugly if we're not careful.

Comment (cvennel) [reply Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: The other option is to write a method that checks the cast. The accepted answer here was a pretty good explanation https://stackoverflow.com/questions/36569421/kotlin-how-to-work-with-list-casts-unchecked-cast-kotlin-collections-listkot

Comment (cvennel) [reply Simulator/src/main/kotlin/edu/drexel/se575/contract/Interpreter.kt]: Or just suppress it

**Pull request merged**